source/fundamentals/gridfs.txt

```diff
@@ -4,7 +4,369 @@ GridFS
+custom name, call the ``SetName()`` method on a ``BucketOptions`` instance as 
+follows:
+
+.. code-block:: go
+
+   db := client.Database("myDB")
+   opts := options.GridFSBucket().SetName("custom name")
+   bucket, err := gridfs.NewBucket(db, opts)
+   
+   if err != nil {
+      panic(err)
+   }
+
+.. _golang-upload-files:
+
+Upload Files
+~~~~~~~~~~~~
+
+You can upload a file into a GridFS bucket in one of the following ways:
+
+- Use the ``UploadFromStream()`` method, which reads from an input stream.
+- Use the ``OpenUploadStream()`` method, which writes to an output stream.
+
+For either upload process, you can specify configuration information on an instance
+of ``UploadOptions``. For a full list of ``UploadOptions`` fields, visit the 
+`API documentation <{+api+}/mongo/options#UploadOptions>`__.
+
+Upload with an Input Stream
+```````````````````````````
+
+To upload a file with an input stream, use the ``UploadFromStream()`` method
+with the following parameters:
+
+- Your file name
+- An ``io.Reader``, with your opened file as a parameter
+- An optional ``opts`` parameter to modify the behavior of ``UploadFromStream()``
+
+The following code example reads from a file called ``file.txt`` and uploads the
+content to a GridFS bucket. It uses an ``opts`` parameter to set file metadata:
+
+.. io-code-block::
+   :copyable: true
+
+   .. input::
+      :language: go
+
+      fileContent, err := os.Open("path/to/file.txt")
+      uploadOpts := options.GridFSUpload().SetMetadata(bson.D{{"metadata tag", "first"}})
+
+      objectID, err := bucket.UploadFromStream("file.txt", io.Reader(fileContent),
+         uploadOpts)
+      if err != nil {
+         panic(err)
+      }
+
+      fmt.Printf("New file uploaded with ID %s", objectID)
+
+   .. output::
+      :language: none
+      :visible: false
+
+      New file uploaded with ID 62e005408bc04f3816b8bcd2
+
+.. note::
+
+   The driver uniquely generates each object ID number. The ID number outputted
+   will resemble the sample output but varies with each file and user.
+
+Upload with an Output Stream
+````````````````````````````
+
+To upload a file with an output stream, use the ``OpenUploadStream()`` method
+with the following parameters:
+
+- Your file name
+- An optional ``opts`` parameter to modify the behavior of ``OpenUploadStream()``
+
+The following code example opens an upload stream on a GridFS bucket and sets
+the number of bytes in each chunk with an ``opts`` parameter. Then, it calls
+the ``Write()`` method on the content of ``file.txt`` to write its content to
+the stream:
+
+.. io-code-block::
+   :copyable: true
+
+   .. input::
+      :language: go
+
+      fileContent, err := os.Open("path/to/file.txt")
+      uploadOpts := options.GridFSUpload().SetChunkSizeBytes(200000)
+      uploadStream, err := bucket.OpenUploadStream("file.txt", uploadOpts)
+
+      if bytes, err = uploadStream.Write(fileContent); err != nil {
```

Comment:
This line does not compile:

```
docs/main.go:69:5: undefined: bytes
docs/main.go:69:16: cannot assign int to bytes in multiple assignment
docs/main.go:69:36: cannot use fileContent (type *os.File) as type []byte in argument to uploadStream.Write
```

Suggest putting these code examples in a `.go` file in the `source/includes/fundamentals/code-snippets` directory. The snippets can be included with `begin` and `end` comments. That may make it easier to validate these examples are correct when updating.

Here is an example:

```go
// source/includes/fundamentals/code-snippets/gridfs.go
package main

import (
	"context"
	"fmt"
	"io"
	"os"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/gridfs"
	"go.mongodb.org/mongo-driver/mongo/options"
)

func main() {
	// Create a new client and connect to the server
	client, err := mongo.Connect(context.TODO())

	if err != nil {
		panic(err)
	}
	defer client.Disconnect(context.TODO())

	bucket, err := gridfs.NewBucket(client.Database("foo"))
	if err != nil {
		panic(err)
	}

	// begin OpenUploadStream example
	file, err := os.Open("go.mod")
	if err != nil {
		panic(err)
	}
	uploadOpts := options.GridFSUpload().SetChunkSizeBytes(200000)
	uploadStream, err := bucket.OpenUploadStream("file.txt", uploadOpts)
	if err != nil {
		panic(err)
	}
	fileContent, err := io.ReadAll(file)
	if err != nil {
		panic(err)
	}
	var bytes int
	if bytes, err = uploadStream.Write(fileContent); err != nil {
		panic(err)
	}
	fmt.Printf("New file uploaded with %d bytes written", bytes)
	// end OpenUploadStream example
}
```

Reply:
I created a `gridfs.go` file in code-snippets and pasted this code there.